README
= jsvCard - JavaScript vCard reader and parser

The main goal of this library is to make the tedious 
personal data forms filling easier using your own vCard.

To do this a parsing library has been developed from scratch
and, taking advantage of HTML5 (jQuery + php when not available),
it is possible to read the vCard directly from filesystem.

== Requirements

jsvCard works within HTML5 browsers (TODO:list of tested browsers)
and has fallback for older browsers using a php stub 
(TODO:state the version of PHP needeD)

== Usage

If you checked out from repository you should run through the
building cycle. See BUILDING for the details.

For a running implementation use test.html. If your browser
does not support HTML5 File API you should configure your
environment to use freadstub.php.

Configuration procedure:

1. Place the content of the "out" folder in a reachable place
2. Place "freadstub.php" in a reachable and phpable place
2. Link the jquery, ajaxuploader and jsvcard (in this order) in
your target webpage
3. Put a clickable element (an image, a button, ...) with a
specified id (say "vClickMe")
4. Create your form and put a specified id (say "vTargetForm")
5. Insert the right classes to the form elements to have them 
populated with the vcard content (see below)
5. Put the init script right before the </body>:

<script type="text/javascript">
//<![CDATA[
	jsvcard = new jsvCard();
	jsvcard.install("vTargetForm","vClickMe","bottom-left");
//]]>
</script>

Note: You can install jsvcard on any pair of "clickable",
"target form" in the same page. DON'T reuse clickables or forms. 

Note: You can skip linking ajaxuploader and installing 
freadstub.php if you don't want support for HTML4. You can 
specify the path for freadstub.php writing this line before the 
"install" call:

	jsvcard.freadstub = "http://www.my-domain.net/freadstub.php";

Note: jsvcard* identifiers and classes are RESERVED, so use other names.

== Supported positions

The position tells how to place the popup respect to the "vClickMe" 
element.

Positions:
* top
* top-left
* top-right
* left
* right
* bottom
* bottom-left
* bottom-right

== Supported classes

TODO: write the supported classes based on fillForm()

== I18n

Messages are put in some variables inside the jsvCard object. 
You can change the messages changing the content of these variables.
Variables names.
* chooseText
* dropText

Pleas look at the source code for a precise reference about these 
variables, their content ahd thir meaning.


CHANGELOG
= jsvCard Changes

== TODO

* vCard 3.0 support
* meaningful unit tests
* close button in popup

== Release 0.1

First release.
* vCard 2.1 supported
* Popup with file selector, url or drop area
* Browser detection for HTML5 support
* Form filling with the vCard content


test/vcard.test.js
describe('Method findElement()', function(){
   vc = new vCard();
	
	it('should find the first element with that elementName case insensitive', function(){
		var data = [{ name:'beGin', value:'vcard'},{name:'begin',value:'vscars'}];
		var index = vc.findElement(data,'begin');
		
		expect(index).not.toEqual(-1);
		expect(data[index]).toEqual({ name:'beGin', value:'vcard'});
	});
	
	it('should find the first element with that elementName and that elementValue case insensitive', function(){
		var data = [{ name:'beGin', value:'vcard'},{name:'begin',value:'otheR'},{name:'begin',value:'oTHeR'}];
		var index = vc.findElement(data,'begin','other');
		
		expect(index).not.toEqual(-1);
		expect(data[index]).toEqual({name:'begin',value:'otheR'});
	});

	it('should find the first element with that elementName and that elementValue = null case insensitive', function(){
		var data = [{ name:'beGin', value:'vcard'},{name:'begin',value:'otheR'},{name:'begin',value:null}];
		var index = vc.findElement(data,'begin',null);
		
		expect(index).not.toEqual(-1);
		expect(data[index]).toEqual({name:'begin',value:null});
	});
		
	it('should not find non-existing elements', function(){
		var data = [{ name:'beGin', value:'vcard'},{name:'begin',value:'otheR'},{name:'begin',value:null}];
		var index = vc.findElement(data,'pip','pap');
		
		expect(index).toEqual(-1);
	});

});


src/base64.js
/* Copyright (C) 1999 Masanao Izumo <iz@onicos.co.jp>
* Version: 1.0
* LastModified: Dec 25 1999
* This library is free.  You can redistribute it and/or modify it.
*/

/**
* Slightly modified respect to the original.
* Defined the lib as a singleton
*/

var Base64 = {
        base64EncodeChars: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/",

        base64DecodeChars: [
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63,
            52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1, -1,
            -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
            15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1,
            -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
            41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, -1, -1, -1, -1, -1 ],

    encode: function (str) {
            var out, i, len;
            var c1, c2, c3;
        
            len = str.length;
            i = 0;
            out = "";
            while(i < len) {
            c1 = str.charCodeAt(i++) & 0xff;
            if(i == len)
            {
                out += base64EncodeChars.charAt(c1 >> 2);
                out += base64EncodeChars.charAt((c1 & 0x3) << 4);
                out += "==";
                break;
            }
            c2 = str.charCodeAt(i++);
            if(i == len)
            {
                out += base64EncodeChars.charAt(c1 >> 2);
                out += base64EncodeChars.charAt(((c1 & 0x3)<< 4) | ((c2 & 0xF0) >> 4));
                out += base64EncodeChars.charAt((c2 & 0xF) << 2);
                out += "=";
                break;
            }
            c3 = str.charCodeAt(i++);
            out += base64EncodeChars.charAt(c1 >> 2);
            out += base64EncodeChars.charAt(((c1 & 0x3)<< 4) | ((c2 & 0xF0) >> 4));
            out += base64EncodeChars.charAt(((c2 & 0xF) << 2) | ((c3 & 0xC0) >>6));
            out += base64EncodeChars.charAt(c3 & 0x3F);
            }
            return out;
    },

    decode: function (str) {
            var c1, c2, c3, c4;
            var i, len, out;
        
            len = str.length;
            i = 0;
            out = "";
            while(i < len) {
                /* c1 */
                do {
                    c1 = base64DecodeChars[str.charCodeAt(i++) & 0xff];
                } while(i < len && c1 == -1);
                if(c1 == -1){
                    break;
                }
                /* c2 */
                do {
                    c2 = base64DecodeChars[str.charCodeAt(i++) & 0xff];
                } while(i < len && c2 == -1);
                if(c2 == -1){
                    break;
                }
                out += String.fromCharCode((c1 << 2) | ((c2 & 0x30) >> 4));
            
                /* c3 */
                do {
                    c3 = str.charCodeAt(i++) & 0xff;
                    if(c3 == 61){
                        return out;
                    }
                    c3 = base64DecodeChars[c3];
                } while(i < len && c3 == -1);
                if(c3 == -1){
                    break;
                }
                out += String.fromCharCode(((c2 & 0XF) << 4) | ((c3 & 0x3C) >> 2));
            
                /* c4 */
                do {
                    c4 = str.charCodeAt(i++) & 0xff;
                    if(c4 == 61){
                        return out;
                    }
                    c4 = base64DecodeChars[c4];
                } while(i < len && c4 == -1);
                if(c4 == -1) {
                    break;
                }
                out += String.fromCharCode(((c3 & 0x03) << 6) | c4);
            }
            return out;
        }
};

src/freadstub.php
<?php
/*
 * @package jsvCard
 * @author Marco Cosentino
 * @copyright (C) 2011 www.professioneit.com
 * @license GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
*/

/*
 * This php stub receives a file and stores it in session. Afterward the file can be read 
 * issuing a get request.
 * 
 * The form submitted must contain a 'element' field which contains the file element name
 * we have in the same form.
 * 
 * Please note that reading through a get request destroys the stored data. Not very restful
 * but this stub is designed to be used in single put & get lifecycles for small contents.
 */

$SID = session_id();
if(empty($SID)) session_start() or exit(basename(__FILE__).': Could not start session');

if( isset($_REQUEST) && isset($_REQUEST['element']) )
	$elem = $_REQUEST['element'];
	
// Implicit put
if( empty($_REQUEST['action']) || $_REQUEST['action'] == 'put')
{
	if( empty($elem) ){
		header("Status: 500 Error in the upload phase -- Missing 'element'");
		exit();
	}
	header('Content-length: 0');
	
	unset($_SESSION['vcardfile']);
	unset($_SESSION['vcardfileerror']);
	
	if(isset($_FILES[$elem]))
	{
		if ($_FILES[$elem]["error"] > 0 /*|| !is_uploaded_file($_FILES["vcardfile"]["error"])*/)
		{
			$_SESSION['vcardfileerror'] = $_FILES[$elem]["error"];
		}
		else
		{
			$_SESSION['vcardfile'] = '';
		
			$handle = fopen($_FILES[$elem]["tmp_name"],'rb');
			while (!feof($handle)) {
		  		$_SESSION['vcardfile'] .= fread($handle, 8192);
			}
		}
	}
	else $_SESSION['vcardfileerror'] = 'File not sent';
}

// The get must be explicit
if(isset($_REQUEST['action']) && $_REQUEST['action'] == 'get')
{

	if( isset($_SESSION['vcardfileerror']) )
	{
		if($_SESSION['vcardfileerror'] == 4)
			header("Status: 599 Error in the upload phase -- Missing file");
		else 
			header("Status: 500 Error in the upload phase -- ".$_SESSION['vcardfileerror']);
			
		$_SESSION = array();
		session_destroy();
	}
	else if( isset($_SESSION['vcardfile']) )
	{
		header('Content-Type: plain/text');
		header('Content-Transfer-Encoding: 8bit');
		header('Content-Length: '.strlen($_SESSION['vcardfile']));
	
		echo $_SESSION['vcardfile'];
		$_SESSION = array();
		session_destroy();
	}
	else
	{
		header("Status: 500 Error in the download phase -- No valid session to read.");
	}
}




src/jsvcard.js
/*
 * @package jsvCard
 * @author Marco Cosentino
 * @copyright (C) 2011 www.professioneit.com
 * @license GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
*/

/**
 * This module contains the jsvCard implementation: a vCard parser and
 * a form filler all made in HTML5 (HTML4 compatible) + jQuery. 
 * @module jsvCard
 * @requires jQuery, jsIconv
 */

/**
 * This is the main jsvCard object you should instantiate this 
 * only once in your web page.
 * @class jsvCard
 * @constructor
 */
var jsvCard = function(){
        
        var self = this;
        
        /**
         * This property sets the path to the freadstub, a stub used to read
         * the file without HTML5
         * @property freadstub
         */
        this.freadstub = "freadstub.php";
        
        this.chooseText = "Please choose the vCard file into the appropriate box.";
        this.dropText = "Alternatively you can drop your vCard file in the drop area below.";
        this.dropHereText = "Drop Here!";
        this.wrongInputText = "Please specify a vCard file in your filesystem.";
        this.wrongDropText = "Please drop only a valid file in the drop area.";
        this.errorReadingText = "An error occoured while reading your vCard file.";
        this.resetText = "Form Reset";
        
        this.forceHTML4 = false;
        
        /**
        * This function installs jsvcard on the specified form
        * and associate the popup to the click event of the clickable element
        * @method install
        */
        this.install = function(form_id,clickable_id,position_mode){
                
                // How many divs are already installed ?
                
                var installed = $("div.jsvcard").lenght !== undefined ? $("div.jsvcard").lenght : 0;
                var current_div = 'jsvcard' + installed;
                var current_div_canvas = 'jsvcardouter' + installed;
                var current_file = 'jsvcardfile' + installed;
                var current_url = 'jsvcardurl' + installed;
                var current_drop = 'jsvcarddrop' + installed;
                var current_reset = 'jsvcardreset' + installed;
                
                $('<div id="'+current_div_canvas+'" class="jsvcardouter"><div id="'+current_div+'" class="jsvcard"><p>'+this.chooseText+'</p></div></div>').hide().appendTo('body');
                
                $("#"+clickable_id).click(function(event){
                        $("#"+current_div_canvas).toggle();
                });
                
                // Position the div into the page
                // when the DOM is ready
                var calcpos_lambda = function(){
                        var pos = self.calculatePosition(
                                        $("#"+clickable_id).position(),
                                        {width:$("#"+clickable_id).width(), height:$("#"+clickable_id).height()},
                                        {width:336, height:376},
                                        position_mode);
                        $("#"+current_div_canvas).css({top:pos.top+"px",left:pos.left+"px"});			
                };
                $("#"+clickable_id).load(calcpos_lambda);
                $(document).ready(calcpos_lambda);
                $(window).resize(calcpos_lambda);

                $('<input type="file" id="'+current_file+'" name="'+current_file+'" single="single" size="15"/><br><br><input id="'+current_reset+'" type="button" value="'+this.resetText+'"/>').appendTo("#"+current_div);
                
                var params = {
                        html5:false,
                        form:form_id,
                        file:current_file,
                        url:current_url,
                        div_to_hide: current_div_canvas
                };
                
                // Install event handlers
                $('#'+current_reset).click(params,this.handleReset);

                // Check for the various File API support.
                if (window.File && window.FileReader && window.FileList && window.Blob && !this.forceHTML4) {
                        
                        // Using HTML5 File API	
                        params.html5 = true;
                        $('#'+current_file).change(params, this.handleLoad);
                        
                        // Determine if the drop event is supported
                        var dnd = "";
                        if(this.isEventSupported('dragstart') && this.isEventSupported('drop')){
                                
                                // Drag & drop supported
                                $("#"+current_div_canvas).addClass('dropok');
                                $("#"+current_div).addClass('dropok');
                                
                                dnd = '<p>'+this.dropText+'</p>'+'<div id="'+current_drop+'" class="jsvcarddrop">'+this.dropHereText+'</div>';
                                $(dnd).appendTo("#"+current_div);
                                
                                $('#'+current_drop).bind('dragover', this.handleDragOver);
                                $('#'+current_drop).bind('drop', params, this.handleDrop);
                        }
                        testdiv = null;
                } else {
                    //Using freadstub
                    $('#'+current_file).change(params, this.handleLoad);
                }
                
        };
        
        /**
         * This function is used to calculate the position of the popup
         * @method calculatePosition
         */
        this.calculatePosition = function(clpos,clsize,mysize,mode){
                var half_mysize = {width:mysize.width/2, height:mysize.height/2};
                var half_clsize = {width:clsize.width/2, height:clsize.height/2};
                var ret = {top:0, left:0};
                switch(mode){
                case "top":
                        ret.top = clpos.top - mysize.height;
                        ret.left = clpos.left + half_clsize.width - half_mysize.width;
                        break;
                case "top-left":
                        break;
                case "top-right":
                        break;
                case "bottom-right":
                        ret.top = clpos.top + clsize.height;
                        ret.left = clpos.left + clsize.width;
                        break;
                }
                return ret;
        };
        
        /**
        * Handles the normal file loading
        * @method handleLoad
        */
        this.handleLoad = function(evt){
                var files = document.getElementById(evt.data.file).files;
                
                if(evt.data.html5 && files !== undefined && files[0] !== undefined){
                    self.loadFromFile(files[0],evt.data.form);
                } else {
                    self.loadFromFileNoHTML5(evt.data.file,evt.data.form);
                }
                $('#'+evt.data.div_to_hide).hide();
                return true;
        };
        
        this.handleReset = function(evt){
                self.resetForm(evt.data.form);
                $('#'+evt.data.div_to_hide).hide();
        };
        
        this.handleDrop = function(evt){
            evt.stopPropagation();
            evt.preventDefault();
                
                var files = evt.originalEvent.dataTransfer.files; 
                
                // Take just the first file
                if(files[0]){
                        self.loadFromFile(files[0],evt.data.form);
                }
                else{
                        alert(self.wrongDropText);
                }
                $('#'+evt.data.div_to_hide).hide();
        };

        this.handleDragOver = function(evt) {
            evt.stopPropagation();
            evt.preventDefault();
        };
        
        
    /*
    * Helpers
    */
    
    /*
    * Loads the vCard from a file using the File API (checked before calling)
    * file - a File object that represents the selected file to load
    * form - a String with the ID of the form to fill with data
    */
    this.loadFromFile = function(file,form){
        var reader = new FileReader();
        
        reader.onload = function(e) {
                self.parseAndFill(e.target.result,form);
        };
        reader.readAsBinaryString(file);
    };
    
    /*
    * Loads the vCard from a file using the freadstub method.
    * The existence of $.ajaxFileUpload must be checked before calling
    * file - a String with the ID of the input element
    * form - a String with the ID of the form to fill with data
    */
    this.loadFromFileNoHTML5 = function(file,form){
        
        // Create a hidden iframe that will receive the content of the submission
        var random = new Date().getTime();
        
        //var src = window.ActiveXObject ? (' src="' + 'javascript:false' + '"') : "";
        
        var ifr_id = 'jsvcardiframe'+ random;
        var ifr = $('<iframe id="'+ifr_id+'" name="'+ifr_id+'"></iframe>').hide().appendTo('body');
        
        var iform_id = 'jsvcardform'+ random;
        var iform = $('<form id="' + iform_id + '" name="' + iform_id + '" action="" method="post" target="" enctype="multipart/form-data"></form>').hide().appendTo('body');
        
        // Clone the file upload element and put the original one in the hidden form
        // The cloned one will replace the original
        var file_clone = $('#'+file).clone();
        $(file_clone).attr('id',"temp"+random);
        $(file_clone).insertBefore('#'+file);
        $('#'+file).appendTo(iform);
        $('<input type="hidden" name="element"/>').val(file).appendTo(iform);
        
        var requestDone = false;
        var timer = null;
        
        // Callback used to load the 
        var loadCallbackLambda = function(isTimeout){
            requestDone = true;
            if(timer){
                    clearTimeout(timer);
            }
            try{
                setTimeout( function(){
                        $(ifr).remove();
                        $(iform).remove();
                }, 100);
            } 
            catch(e){}
            
            if(isTimeout != "timeout"){
                self.freadstubPutDone(form);
            }
            else{
                alert(self.errorReadingText + "\nRequest timeout!");
            }
            // Here we restore the id for the input
                $('#'+"temp"+random).attr('id',file);
        };
        
        $(ifr).load(loadCallbackLambda);
        
        $(iform).attr('action', self.freadstub);
        $(iform).attr('method', 'post');
        $(iform).attr('target', ifr_id);
        if(iform.encoding){
            iform.encoding = 'multipart/form-data';
        }
        else{
                iform.enctype = 'multipart/form-data';
        }
        
        try{
            $(iform).submit();
            timer = setTimeout(function(){
                // Check to see if the request is still happening
                if( !requestDone ) {
                    loadCallbackLambda( "timeout" );
                }
            }, 10000); // 10 secs.
        } 
        catch(e){
                alert(self.errorReadingText + "\n" + e);
        }
    };
    
    this.freadstubPutDone = function(form){
        $.ajax({
                url: self.freadstub,
                data: {action: 'get'},
                success: function(data) {
                        self.parseAndFill(data,form);
                },
                statusCode:{
                        599: function(){
                                alert(self.errorReadingText + "\nFreadstub reported: Missing content!");
                        },
                        500: function(){
                                alert(self.errorReadingText + "\nFreadstub reported: Generic error!");
                        }
                }
        });
    };
    

    this.parseAndFill = function(data,formid){
        var vcard = new vCard(); 
        var vcdata = vcard.parsevCardData(data);
        self.fillForm('vTargetForm',vcdata);
    };
    
    
	/**
	 * This hash contains classes associations. The key is the classname
	 * the value is the path within the vcard object structure.<br/>
	 * This hash will be used to fill the target form. If you want to support
	 * other classes you have to add items to this hash [TODO: write how to]
	 * @property CLASSES
	 * @type Object
	 */
	this.CLASSES = {
    		vVersion:"version",
    		vTitle:"title",
    		vFirstName:"n.first",
    		vMidNames:"n.middle",
    		vLastName:"n.last",
    		vNamePrefix:"n.prefix",
    		vNamePostfix:"n.postfix",
    		vFName:"fn",
    		vAddressHomePO:"adr.home.po",
    		vAddressHomeExt:"adr.home.extension",
    		vAddressHomeStreet:"adr.home.street",
    		vAddressHomeLocation:"adr.home.location",
    		vAddressHomeRegion:"adr.home.region",
    		vAddressHomePostalCode:"adr.home.postalcode",
    		vAddressHomeCountry:"adr.home.country",
    		vAddressHomeLabel:"label.home",
    		vTelHomeVoice:"tel.home.voice",
    		vTelHomeFax:"tel.home.fax",
    		vTelWorkVoice:"tel.work.voice",
    		vTelWorkFax:"tel.work.fax"    		
    };

    /**
    * Fills the specified form with vCard parsed data.
    * If you don't like the class names we use you can override this
    * function with your own.
    * @method fillForm
    * formid - a String with the ID of the form to fill with data
    * vcard - an Object that holds the data
    */
	this.fillForm = function(formid,vcard)
    {
		// Cycle through classes
		for (var klass in self.CLASSES){
			if (self.CLASSES.hasOwnProperty(klass)) {
			
				var splitted = self.CLASSES[klass].split(".");
				var i = 0;
				var current_obj = vcard;
				
				// Check if the specified path is defined inside the vcard object
				while(i < splitted.length && current_obj !== undefined && current_obj.hasOwnProperty(splitted[i])){
					current_obj = current_obj[splitted[i]];
					i++;
				}
				
				// If the whole path inside the vcard object is defined then we
				// will have the final value inside current_obj
				if(i == splitted.length){
					var element = $('form#'+formid+' .'+klass);
					if(element[0] !== undefined){
						switch(element[0].nodeName.toLowerCase())
						{
						case 'input':
							$(element).val(current_obj);
							break;
						case 'textarea':
							$(element).html(current_obj);
							break;
						}
					}
				}
			}
		}
    };
        
    this.resetForm = function(formid){
            $('form#'+formid+' input').val('');
            $('form#'+formid+' textarea').html('');
    };
                    
	/**
	* Determines if a given element supports the given event
	* function from http://yura.thinkweb2.com/isEventSupported/
	* @method isEventSupported
	*/
    this.isEventSupported = function ( eventName, element ) {

        var TAGNAMES = {
                    'select': 'input', 'change': 'input',
                    'submit': 'form', 'reset': 'form',
                    'error': 'img', 'load': 'img', 'abort': 'img'
                };

	    element = element || document.createElement(TAGNAMES[eventName] || 'div');
	    eventName = 'on' + eventName;
	
	    // When using `setAttribute`, IE skips "unload", WebKit skips "unload" and "resize", whereas `in` "catches" those
	    var isSupported = eventName in element;
	
	    if ( !isSupported ) {
	        // If it has no `setAttribute` (i.e. doesn't implement Node interface), try generic element
	        if ( !element.setAttribute ) {
	            element = document.createElement('div');
	        }
	        if ( element.setAttribute && element.removeAttribute ) {
	            element.setAttribute(eventName, '');
	            isSupported = is(element[eventName], 'function');
	
	            // If property was created, "remove it" (by setting value to `undefined`)
	            if ( !is(element[eventName], 'undefined') ) { element[eventName] = undefined; }
	            element.removeAttribute(eventName);
	        }
	    }
	
	    element = null;
	    return isSupported;
    };
    
};

